src/Simulation/Simulators/OpenSystemsSimulator/DataModel/ArrayConverter.cs

```diff
@@ -3,249 +3,245 @@
+                    reader.Read();
+                    version = reader.GetInt32();
                     break;
-                }
 
-                // If it's not the end of the object, the current token needs
-                // to be a property name.
-                if (reader.TokenType != JsonTokenType.PropertyName)
-                {
-                    throw new JsonException();
-                }
+                case "dim":
+                    dims = new List<int>();
+                    reader.Require(JsonTokenType.StartArray);
+                    while (reader.Read() && reader.TokenType != JsonTokenType.EndArray)
+                    {
+                        dims.Add(reader.GetInt32());
+                    }
+                    break;
 
-                var propertyName = reader.GetString();
+                case "data":
+                    data = new List<Complex>();
+                    reader.Require(JsonTokenType.StartArray);
+
+                    while (reader.Read() && reader.TokenType != JsonTokenType.EndArray)
+                    {
+                        // We expect an inner array at this point, so
+                        // go on and read the begin array token.
+                        reader.Require(JsonTokenType.StartArray, read: false);
 
-                switch (propertyName)
-                {
-                    case "v":
                         reader.Read();
-                        version = reader.GetInt32();
-                        break;
-
-                    case "dim":
-                        dims = new List<int>();
-                        reader.Require(JsonTokenType.StartArray);
-                        while (reader.Read() && reader.TokenType != JsonTokenType.EndArray)
-                        {
-                            dims.Add(reader.GetInt32());
-                        }
-                        break;
-
-                    case "data":
-                        data = new List<Complex>();
-                        reader.Require(JsonTokenType.StartArray);
-
-                        while (reader.Read() && reader.TokenType != JsonTokenType.EndArray)
-                        {
-                            // We expect an inner array at this point, so
-                            // go on and read the begin array token.
-                            reader.Require(JsonTokenType.StartArray, read: false);
-
-                            reader.Read();
-                            var real = reader.GetDouble();
-
-                            reader.Read();
-                            var imag = reader.GetDouble();
-
-                            reader.Require(JsonTokenType.EndArray);
-
-                            data.Add(new Complex(real, imag));
-                        }
-                        break;
-
-                    default:
-                        throw new JsonException();
-                }
-            }
+                        var real = reader.GetDouble();
```

Comment:
(Paranoid)
In this line I feel a risk of interpreting the possible `JsonTokenType.EndArray` (immediately following the `JsonTokenType.StartArray` in line 98) as `double`.

Up to you to decide whether that is possible or not.

Reply:
Same for L104.

Reply:
The `GetDouble` method throws an `InvalidOperationException` when the token type is not a JSON number such that if the array contains the wrong number of elements (an array representing a complex should always have length exactly two), that gets propagated out into an exception. It could be safer to write a new method that wraps that in a more actionable error message, but I think that's likely overkill here?